test/integration/upstreams/per_host_upstream_request.h

```diff
@@ -0,0 +1,111 @@
+      }
+    }
+  }
+}
+} // namespace
+class PerHostHttpConnPool : public Router::GenericConnPool,
+                            public Envoy::Http::ConnectionPool::Callbacks {
+public:
+  // GenericConnPool
+  PerHostHttpConnPool(Upstream::ClusterManager& cm, bool is_connect,
+                      const Router::RouteEntry& route_entry,
+                      absl::optional<Envoy::Http::Protocol> downstream_protocol,
+                      Upstream::LoadBalancerContext* ctx) {
+    ASSERT(!is_connect);
+    conn_pool_ = cm.httpConnPoolForCluster(route_entry.clusterName(), route_entry.priority(),
+                                           downstream_protocol, ctx);
+  }
+  void newStream(Router::GenericConnectionPoolCallbacks* callbacks) override;
+  bool cancelAnyPendingStream() override;
+  absl::optional<Envoy::Http::Protocol> protocol() const override;
+
+  // Http::ConnectionPool::Callbacks
+  void onPoolFailure(ConnectionPool::PoolFailureReason reason,
+                     absl::string_view transport_failure_reason,
+                     Upstream::HostDescriptionConstSharedPtr host) override;
+  void onPoolReady(Envoy::Http::RequestEncoder& callbacks_encoder,
+                   Upstream::HostDescriptionConstSharedPtr host,
+                   const StreamInfo::StreamInfo& info) override;
+  Upstream::HostDescriptionConstSharedPtr host() const override { return conn_pool_->host(); }
+
+  bool valid() { return conn_pool_ != nullptr; }
+
+private:
+  // Points to the actual connection pool to create streams from.
+  Envoy::Http::ConnectionPool::Instance* conn_pool_{};
+  // The handle to the upstream request. Non-null if there is pending stream to create.
+  Envoy::Http::ConnectionPool::Cancellable* conn_pool_stream_handle_{};
+  Router::GenericConnectionPoolCallbacks* callbacks_{};
+};
+
+class PerHostHttpUpstream : public Router::GenericUpstream {
+public:
+  PerHostHttpUpstream(Router::UpstreamToDownstream& upstream_request,
+                      Envoy::Http::RequestEncoder* encoder,
+                      Upstream::HostDescriptionConstSharedPtr host)
+      : sub_upstream_(upstream_request, encoder), host_(host) {}
+
+  // GenericUpstream
+  void encodeData(Buffer::Instance& data, bool end_stream) override {
+    sub_upstream_.encodeData(data, end_stream);
+  }
+
+  void encodeMetadata(const Envoy::Http::MetadataMapVector& metadata_map_vector) override {
+    sub_upstream_.encodeMetadata(metadata_map_vector);
+  }
+
+  Http::Status encodeHeaders(const Envoy::Http::RequestHeaderMap& headers,
+                             bool end_stream) override {
+    auto dup = Envoy::Http::RequestHeaderMapImpl::create();
+    Envoy::Http::HeaderMapImpl::copyFrom(*dup, headers);
+    dup->setCopy(Envoy::Http::LowerCaseString("X-foo"), "foo-common");
+    addHeader(*dup, "X-cluster-foo", host_->cluster().metadata(), "foo", "bar");
+    if (host_->metadata() != nullptr) {
+      addHeader(*dup, "X-host-foo", *host_->metadata(), "foo", "bar");
+    }
```

Comment:
If you're only trying to verify that you can modify headers in the upstream extension why the complexity here? Why not just a single header?

Reply:
The idea is that I don't want to add test case because 
"Envoy can do this in this current implementation" but  
"This is real world requirement"
Also this can acts an example for user who want to rewrite header base on per upstream attribute like #13897

Does it make sense?